### PR TITLE
Do not emit an event if it will cause a loop

### PR DESCRIPTION
**Pull request by Pchelolo**
After [an outage](https://phabricator.wikimedia.org/T134537) we've discovered, that when change-prop is introduced, some bugs playing together might cause rerender loops in RESTBase, when change-prop serenaders some content, and it creates the same event that caused a rerender to happen. 

This PR adds some protection against it - change-prop will add the `x-rerender-reason` header (name subject to a bike shed) to a RESTBase request, containing an original event URI (it's enough as RB only emits `resource_change` events). If we are about to emit the same event causing a cycle - log an error and drop the event.

Questions:
1. What to use as an event hash. I think URI is enough, but suggestions are welcome
2. This doesn't prevent multiple-step loops (event1 causes rerender1 causes event2 causes rerender2 causes event1). Do we want to support cases like that? As a generic way of preventing multi-step loops, we've had an idea to add something like an XFF header to the event meta to indicate the whole sequence of events that caused some event to fire. This could be useful for debugging as well as to prevent multi-step loops

PS: need to add this header in change-propagation.

cc @wikimedia/services

### CODE REVIEW

**Comment by gwicke** [comment]
> 1. What to use as an event hash. I think URI is enough, but suggestions are welcome

Perhaps `{topic}:{url}`, as in `resource_change://en.wikipedia.org/v1/page/html/Foobar` ?

> 2. This doesn't prevent multiple-step loops (event1 causes rerender1 causes event2 causes rerender2 causes event1). Do we want to support cases like that? As a generic way of preventing multi-step loops, we've had an idea to add something like an XFF header to the event meta to indicate the whole sequence of events that caused some event to fire. This could be useful for debugging as well as to prevent multi-step loops

Yeah, I think this would be really useful. Even in the absence of loops, it's really useful to see where an event came from.

File: sys/events.js [hunk 1]
```diff
@@ -85,6 +85,26 @@ EventService.prototype._emit = function(hyper, req) {
                 return event;
             })
             .filter(function(event) { return !!event; });
+
+            // Change-propagation will set up the x-rerender-reason header, indicating
+            // the event which caused the rerender. In case RESTBase is about to emit
+            // the same event, it will cause a rerender loop. So, log an error and skip
+            // an event.
+            var reRenderReason = req.headers && req.headers['x-rerender-reason']
+                || hyper._rootReq && hyper._rootReq['x-rerender-reason'];
```
Comment (gwicke) [inline sys/events.js]: [minor, bikeshed, alternative]: `x-triggers` or `x-events`?

File: sys/events.js [hunk 2]
```diff
@@ -85,6 +85,26 @@ EventService.prototype._emit = function(hyper, req) {
...
+            if (reRenderReason) {
+                events = events.filter(function(event) {
+                    if (reRenderReason.replace(/^https?:/, '') === event.meta.uri.replace(/^https?:/, '')) {
```
Comment (gwicke) [inline sys/events.js]: It would be good to include the topic as well. Perhaps pre-process the trigger / reason header into an object, and then match on that?

Also, why do we need to replace `https:/` here? I thought our internal events are all proto-relative, is that not the case?

File: sys/events.js (hunk 1 above)
Comment (d00rman) [inline sys/events.js]: :+1: for `x-triggered-by`

File: sys/events.js [hunk 3]
```diff
@@ -85,6 +85,26 @@ EventService.prototype._emit = function(hyper, req) {
                 return event;
             })
             .filter(function(event) { return !!event; });
+
+            // Change-propagation will set up the x-rerender-reason header, indicating
+            // the event which caused the rerender. In case RESTBase is about to emit
+            // the same event, it will cause a rerender loop. So, log an error and skip
+            // an event.
+            var reRenderReason = req.headers && req.headers['x-rerender-reason']
```
Comment (d00rman) [inline sys/events.js]: Minor nit: `reRenderReason` looks ... ugh, unnatural to me. `origReason` perhaps?

File: sys/events.js (hunk 2 above)
Comment (d00rman) [inline sys/events.js]: > It would be good to include the topic as well. Perhaps pre-process the trigger / reason header into an object, and then match on that?

I agree. The combo is what makes the event unique.

> Also, why do we need to replace https:/ here? I thought our internal events are all proto-relative, is that not the case?

I'm reading this as making sure it's proto-relative and as such I think it makes for more solid code.`

Comment (Pchelolo) [inline sys/events.js]: So, as a conclusion, we all agree we need to get the optional XFF-like field in meta in all events. 

As for the regex - basically we're making sure that a loose `s` in https doesn't get us in trouble. Events in the extension are emitted the https, while restbase exists them with http.

File: sys/events.js (hunk 1 above)
Comment (Pchelolo) [inline sys/events.js]: `x-restbase-triggered-by`?

File: sys/events.js (hunk 3 above)
Comment (Pchelolo) [inline sys/events.js]: `origEvent`?

File: sys/events.js (hunk 2 above)
Comment (d00rman) [inline sys/events.js]: Yup for XFF-like header, but IMHO this can be a separate PR.

File: sys/events.js (hunk 3 above)
Comment (d00rman) [inline sys/events.js]: :+1:

File: sys/events.js (hunk 1 above)
Comment (d00rman) [inline sys/events.js]: I think we may want to make it independent of RESTBase. We don't want to re-emit an event for the same URI/topic combo regardless of the original source.

File: sys/events.js (hunk 3 above)
Comment (gwicke) [inline sys/events.js]: triggers?

Comment (d00rman) [inline sys/events.js]: Gotta love bike-shedding :) `trigger`(`s`) LGTM :+1:

**Comment by Pchelolo** [comment]
Updated to match the changes in https://github.com/wikimedia/change-propagation/pull/32

File: sys/events.js [hunk 4]
```diff
@@ -85,6 +85,28 @@ EventService.prototype._emit = function(hyper, req) {
                 return event;
             })
             .filter(function(event) { return !!event; });
+
+            // Change-propagation will set up the x-rerender-reason header, indicating
```
Comment (d00rman) [inline sys/events.js]: x-rerender-reason?

File: sys/events.js [hunk 5]
```diff
@@ -85,6 +85,28 @@ EventService.prototype._emit = function(hyper, req) {
...
+            // the event which caused the rerender. In case RESTBase is about to emit
+            // the same event, it will cause a rerender loop. So, log an error and skip
+            // an event.
```
Comment (d00rman) [inline sys/events.js]: s/an/the/

**Comment by d00rman** [comment]
Minor nits, LGTM otherwise

**Comment by Pchelolo** [comment]
@d00rman updated
